ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/PerasCertDB/Impl.hs

```diff
@@ -0,0 +1,297 @@
+  API implementation
+-------------------------------------------------------------------------------}
+
+implCloseDB :: IOLike m => PerasCertDbHandle m blk -> m ()
+implCloseDB (PerasCertDbHandle varState) =
+  atomically (swapTVar varState PerasCertDbClosed) >>= \case
+    PerasCertDbOpen PerasCertDbEnv{pcdbTracer} -> do
+      traceWith pcdbTracer ClosedPerasCertDB
+    -- DB was already closed.
+    PerasCertDbClosed -> pure ()
+
+-- TODO: we will need to update this method with non-trivial validation logic
+-- see https://github.com/tweag/cardano-peras/issues/120
+implAddCert ::
+  ( IOLike m
+  , StandardHash blk
+  ) =>
+  PerasCertDbEnv m blk ->
+  ValidatedPerasCert blk ->
+  m AddPerasCertResult
+implAddCert env cert = do
+  traceWith pcdbTracer $ AddingPerasCert roundNo boostedPt
+  res <- atomically $ do
+    WithFingerprint
+      PerasVolatileCertState
+        { pvcsCerts
+        , pvcsWeightByPoint
+        , pvcsCertsByTicket
+        , pvcsLastTicketNo
+        }
+      fp <-
+      readTVar pcdbVolatileState
+    if Map.member roundNo pvcsCerts
+      then pure PerasCertAlreadyInDB
+      else do
+        let pvcsLastTicketNo' = succ pvcsLastTicketNo
+        writeTVar pcdbVolatileState $
+          WithFingerprint
+            PerasVolatileCertState
+              { pvcsCerts =
+                  Map.insert roundNo cert pvcsCerts
+              , -- Note that the same block might be boosted by multiple points.
+                pvcsWeightByPoint =
+                  addToPerasWeightSnapshot boostedPt (getPerasCertBoost cert) pvcsWeightByPoint
+              , pvcsCertsByTicket =
+                  Map.insert pvcsLastTicketNo' cert pvcsCertsByTicket
+              , pvcsLastTicketNo = pvcsLastTicketNo'
+              }
+            (succ fp)
+        pure AddedPerasCertToDB
+  traceWith pcdbTracer $ case res of
+    AddedPerasCertToDB -> AddedPerasCert roundNo boostedPt
+    PerasCertAlreadyInDB -> IgnoredCertAlreadyInDB roundNo boostedPt
+  pure res
+ where
+  PerasCertDbEnv
+    { pcdbTracer
+    , pcdbVolatileState
+    } = env
+
+  boostedPt = getPerasCertBoostedBlock cert
+  roundNo = getPerasCertRound cert
+
+implGetWeightSnapshot ::
+  IOLike m =>
+  PerasCertDbEnv m blk -> STM m (WithFingerprint (PerasWeightSnapshot blk))
+implGetWeightSnapshot PerasCertDbEnv{pcdbVolatileState} =
+  fmap pvcsWeightByPoint <$> readTVar pcdbVolatileState
+
+implGetCertSnapshot ::
+  IOLike m =>
+  PerasCertDbEnv m blk -> STM m (PerasCertSnapshot blk)
+implGetCertSnapshot PerasCertDbEnv{pcdbVolatileState} =
+  readTVar pcdbVolatileState
+    <&> forgetFingerprint
+    <&> \PerasVolatileCertState
+           { pvcsCerts
+           , pvcsCertsByTicket
+           } ->
+        PerasCertSnapshot
+          { containsCert = \r -> Map.member r pvcsCerts
+          , getCertsAfter = \ticketNo ->
+              let (_, certs) = Map.split ticketNo pvcsCertsByTicket
+               in [(cert, tno) | (tno, cert) <- Map.toAscList certs]
```

Comment:
Why not returning the pair of maps? Do we need them in list form?

Reply:
Yeah, that sounds like a good refactoring; I don't think this is really leaking implementation details.